### ISSUE TITLE
AsyncLocal not safe when using PerLogicalCallContextScopeManagerProvider in ASP.Net full framework

### ISSUE THREAD

**Issue by micklaw** [seesharper/LightInject#386]
Hi,

First off, I'd just like to say thanks for your time and efforts on this project. Kudos isn't often given in OSS, so have a +1 on me.

Recently a colleague of mine, @t11omas raised an issue we both stumbled across while using the PerLogicalCallContextScopeManagerProvider in a ASP.Net application, which you duly resolved. I took a win from that so thanks again.

Though evolving from that another issue has popped up where on larger requests (file uploads, or big content body) the PerLogicalCallContextScopeManagerProvider throws an exception looking for the scope from the LogicalThreadStorage implementation for NET46. It appears that the LogicalThreadStorage uses AsyncLocal to store the scope across the async operations. The problem we came across is, as we are setting up the scope on the Application_BeginRequest method in our Global.asax.cs, it does not appear to flow across async operations, and thus the AsyncLocal instance is null on the next request. 

I'm not sure of the technical details here around AsyncLocal (more reading required), but it appears it is not safe make use of AsyncLocal when being setup via the Global.asax.cs events.

MiniProfiler also seems to have had this issue previously, and basically I have have ripped off the solution they used as it appears to make sense in my initial requirement and resolves any issues I was seeing. I have included this here in a sample project for reference. Here is the snippet I've changed too.
[AsyncLocalIssue.zip](https://github.com/seesharper/LightInject/files/1348814/AsyncLocalIssue.zip)

 ```csharp
/// <summary>
/// Provides storage per logical thread of execution.
/// </summary>
/// <typeparam name="T">The type of the value contained in this <see cref="LogicalThreadStorage{T}"/>.</typeparam>
[System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
internal class LogicalThreadStorage<T>
{
    private readonly AsyncLocal<T> asyncLocal = new AsyncLocal<T>();

    private readonly object lockObject = new object();

    /// <summary>
    /// The last instance of the <see cref="T">item</see>
    /// </summary>
    private T _last;

    /// <summary>
    /// Gets or sets the value for the current logical thread of execution.
    /// </summary>
    /// <value>
    /// The value for the current logical thread of execution.
    /// </value>
    public T Value
    {
        get 
        {
            if (this.asyncLocal.Value != null)
            {
                return this.asyncLocal.Value;
            }

            return _last;
        }
        set { asyncLocal.Value = _last = value; }
    }
}
 ```

I'm not quite proposing this as the solution in a PR yet, but more of a discussion,  but as i say it appears to work for the time being. If you feel like this would work for you, I'd be happy to submit it as a PR.

Cheers

Michael

MiniProfiler/dotnet#173
aspnet/AspNetKatana#31
seesharper/LightInject#383

**Comment by seesharper** [comment]
Hi and thanks a lot for taking the time to investigate on this. Kudos appreciated 👍 
I am just going to jump right to your proposed solution here. 

How would you know that `_last` is your last if you know what I mean. It could be mutated by another request by the time you read it. 

btw, do you know of any way to "force/provoke" ASP.NET to lose the async value?

**Comment by micklaw** [comment]
Hey @seesharper, thanks for getting back to me.

In the sample project I attached I am creating the container in the Global.asax.cs, then on BeginRequest the scope is created, and subsequently disposed on EndRequest. As I understand from my basic testing, the LightInject scope gets assigned to the AsyncLocal value within the LogicalThreadStorage<T> that is created once per web request, so it feels safe enough to assume that if only one LogicalThreadStorage is created per web request I can store the value as a kind of 'last one wins' as a fallback in the scenario that the AsyncLocal instance is null.

It is unclear to me exactly why this is clearing out, but it does seem like as per the linked issue in MiniProfiler and Katana that in the old ASP.Net pipeline should not be used with AsyncLocal as per @davidfowl comment further down the page.

> Why not just always store the context in the httpContext? Don't use async locals when you're in System.Web...

In terms of forcing it, in the sample project I sent on I have put this fix in place, so the form will post, but if you used the current implementation of LogicalThreadStorage<T> that is in 5.1.0 the form on the /about page will bail as it has no LightInject scope. As i can see, if you do a POST with only small body the form post completes successfully, but as soon as you got to any sort of size (e.g file upload), it seems to bail. As i say, I'm not able at this stage to give an exact explanation as to what is going on under the hood in the ASP.Net pipeline, but it does seem to handle async process differently when the request is larger. Which causes the initially set AsyncLocal to be new.

Hope this helps, anything else I can do please let me know.

**Comment by seesharper** [comment]
What a conundrum this is :) . HttpContext.Current does not flow across async/await points (if it did we could just store it there) and AsyncLocal does not (always) work through the begin/end-request pipeline :)

Lets begin with the class  `PerLogicalCallContextScopeManager`

https://github.com/seesharper/LightInject/blob/253f284f0e614ddf36a004226acb0f14d5a85878/src/LightInject/LightInject.cs#L4204-L4207

There is only one instance of the `PerLogicalCallContextScopeManager` and hence there is only one LogicalThreadStorage instance as well which is why I'm afraid that `_last` might be altered by an another request. There might be possible to do a combination of AsyncLocal and HttpContext.Current, but since either can really be trusted, I fear that we are only pushing the problem ahead of us 

I have to read up on the links you provided, but I think the root problem here is that the "old" ASP.NET stack was never really designed to handle async/await properly. :)

Again, I really appreciate your time and effort here and would love to up with a solution that is 100% bulletproof.

**Comment by micklaw** [comment]
Yes, it really is making my head itch this one.

So as you summarised, the LogicalThreadStorage class is created as a singleton and can be overwritten at any point by another users request, so this solution would not fit it seems.

Taking a step back, I suppose my specific problem in a web context for the 'old' pipeline is really just the lifetime of the scope and it not being a singleton. Thinking out loud, do you think potentially having a overload on BeginScope to take an instance of a ScopeManager or Provider could allow for us to manage the scope on a scoped instance more accurately than on the Global.asax.cs registration?

**Comment by davidfowl** [comment]
We just released a new feature to help people deal with the way ASP.NET 4.x enters and exits the pipeline. See the ASP.NET execution step feature https://blogs.msdn.microsoft.com/dotnet/2017/09/13/net-framework-4-7-1-asp-net-and-configuration-features/. Unfortunately it's only available on .NET Framework 4.7.1 😄

**Comment by davidfowl** [comment]
> What a conundrum this is :) . HttpContext.Current does not flow across async/await points (if it did we could just store it there)

It does once the user code doesn't get rid of the sync context

**Comment by seesharper** [comment]
Thanks @davidfowl  for taking the time to comment on this.

> It does once the user code doesn't get rid of the sync context

So , no `ConfigureAwait(false)`, is that what you are saying?

Just for the record, this isn't an issue in AspNetCore, right? AsyncLocal are good there?

**Comment by davidfowl** [comment]
> So , no ConfigureAwait(false), is that what you are saying?

That's correct.

> Just for the record, this isn't an issue in AspNetCore, right? AsyncLocal are good there?

Right, its a single async call chain on ASP.NET Core.

On ASP.NET (not core) today, the pipeline looks more like a zig zag:

```
-> Begin Request
        -> Enter managed code
        <-  Exit managed code
-> Next event 
         -> Enter managed code
         <- Exit managed code
-> End Request
       -> Enter managed code
       <- Exit managed code
```

Between each of those events the only thing that is preserved is the `HttpContext`.

**Comment by seesharper** [comment]
Thanks, @davidfowl for the clarification 👍 

@micklaw Are we good here. Basically what we can conclude from this is that we should use LightInject.Web which handles the scopes using `HttpContext` and stay away from `ConfigureAwait(false)` :)

**Comment by micklaw** [comment]
@seesharper, I think so. It does seem from @davidfowl comment that a fix is in place for 4.7.1 which is good enough for me long term, while altering this particular part of my code to maintain the SyncronizationContext is my best option short term (goodbye ConfigureAwait(false) :)).

In my particular instance we are 90% of the way to moving the solution to aspnetcore anyway, just in the interim it would have been nice to have a fix, but it feels like there is no quick win to be had here.

Thanks again to both of you for your help, it’s been a really interesting and enjoyable dive in to ASP.Net and LighyInject. Good up the good fight fellas.
